### PR TITLE
yaksa: upgrade to latest master

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Memory leaks found with AddressSanitizer have been fixed.

## Expected Impact

Fix memory leaks.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
Warning failure can be ignored, but it looks like there's new yaksa handle leaks with atomic pairtype tests. Need to investigate.

**Comment by raffenet** [comment]
There are 2 issues related to the pairtype handle leaks in this PR.
1. predefined pairtypes are flattened in ch4 active messages when they do not need to be. it should be possible to send pairtype handles to target processes, which can interpret them safely.
2. yaksa increases the reference count of "builtin" types if detected during unflattening (https://github.com/pmodels/yaksa/blob/7bff63498c67b26b9231e870c7c38255186990dc/src/frontend/flatten/yaksa_unflatten.c#L23). however when freeing those unflattened types, and exception is made for pairtypes causing their refcounts not to be decremented, which leads to them being leaked (https://github.com/pmodels/mpich/blob/85efc39caa1325b9e2ee6caa75a7cd44329c1457/src/mpi/datatype/typerep/src/typerep_yaksa_commit.c#L67-L71)

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
test:mpich/ch4/most

File: src/include/mpir_datatype.h [hunk 1]
```diff
@@ -329,7 +329,7 @@ void MPIR_Datatype_get_flattened(MPI_Datatype type, void **flattened, int *flatt
 #define MPIR_Datatype_add_ref_if_not_builtin(datatype_)             \
```
Comment (raffenet) [inline src/include/mpir_datatype.h:329]: I opted not to change the name of the macro in this commit. It is used extensively in the code, so would be a lot of churn, and it's likely the intent was to exempt pairtypes all along.

**Comment by pavanbalaji** [comment]
> 2. yaksa increases the reference count of "builtin" types if detected during unflattening (https://github.com/pmodels/yaksa/blob/7bff63498c67b26b9231e870c7c38255186990dc/src/frontend/flatten/yaksa_unflatten.c#L23). however when freeing those unflattened types in MPICH, an exception is made for pairtypes, causing their refcounts not to be decremented. this is the leak we see in this PR. (https://github.com/pmodels/mpich/blob/85efc39caa1325b9e2ee6caa75a7cd44329c1457/src/mpi/datatype/typerep/src/typerep_yaksa_commit.c#L67-L71
>    )

This PR doesn't fix this part yet, right?

**Comment by raffenet** [comment]
> > 1. yaksa increases the reference count of "builtin" types if detected during unflattening (https://github.com/pmodels/yaksa/blob/7bff63498c67b26b9231e870c7c38255186990dc/src/frontend/flatten/yaksa_unflatten.c#L23). however when freeing those unflattened types in MPICH, an exception is made for pairtypes, causing their refcounts not to be decremented. this is the leak we see in this PR. (https://github.com/pmodels/mpich/blob/85efc39caa1325b9e2ee6caa75a7cd44329c1457/src/mpi/datatype/typerep/src/typerep_yaksa_commit.c#L67-L71
> >    
> >    )
> 
> This PR doesn't fix this part yet, right?

It does. I removed the flattening/unflattening of pairtypes in ch4, so the yaksa refcounts stay constant and are freed correctly in finalize.

**Comment by pavanbalaji** [comment]
I've approved it, but it's probably safe to remove the pair type branches in typerep free

**Comment by raffenet** [comment]
> I've approved it, but it's probably safe to remove the pair type branches in typerep free

I think in order to do that, we need to somehow increase the refcount of the yaksa pairtypes when they are committed. I tried removing the branch, but ran into assertions that pairtype refcounts went negative inside yaksa finalize.
https://github.com/pmodels/mpich/blob/master/src/mpi/datatype/typerep/src/typerep_yaksa_commit.c#L9

**Comment by raffenet** [comment]
Oh, I forgot to test ch3 after adding the MPIR-level changes.
test:mpich/ch3/most

**Pull request merged**